concept/tests/test_thing.rs

```diff
@@ -655,3 +739,203 @@ fn role_player_duplicates() {
+
+        let attr_by_id = thing_manager
+            .get_attribute_with_value(&snapshot, attr_type.clone(), Value::Struct(Cow::Borrowed(&struct_value)))
+            .unwrap()
+            .unwrap();
+        assert_eq!(attr, attr_by_id);
+        assert_eq!(attr_created, attr);
+        snapshot.close_resources();
+    }
+}
+
+#[test]
+fn read_struct_by_field() {
+    let (storage, type_manager, thing_manager) = prepare();
+
+    let attr_label = Label::build("index_test_attr");
+    let nested_struct_def = StructDefinition::define(
+        "nested_test_struct".to_owned(),
+        HashMap::from([("nested_string".to_owned(), (ValueType::String, false))]),
+    );
+
+    let nested_struct_key = {
+        let mut snapshot = storage.clone().open_snapshot_write();
+        let nested_struct_key = type_manager.create_struct(&mut snapshot, nested_struct_def).unwrap();
+        snapshot.commit().unwrap();
+        nested_struct_key
+    };
+
+    let struct_def = StructDefinition::define(
+        "index_test_struct".to_owned(),
+        HashMap::from([("f_nested".to_owned(), (ValueType::Struct(nested_struct_key.clone()), false))]),
+    );
+    let (attr_type, struct_key, struct_def) = {
+        let mut snapshot = storage.clone().open_snapshot_write();
+        let struct_key = type_manager.create_struct(&mut snapshot, struct_def.clone()).unwrap();
+        let mut attr_type = type_manager.create_attribute_type(&mut snapshot, &attr_label, false).unwrap();
+        attr_type.set_value_type(&mut snapshot, &type_manager, ValueType::Struct(struct_key.clone())).unwrap();
+        attr_type
+            .set_annotation(&mut snapshot, &type_manager, AttributeTypeAnnotation::Independent(AnnotationIndependent))
+            .unwrap();
+        snapshot.commit().unwrap();
+        (attr_type, struct_key, struct_def)
+    };
+
+    // Create value
+    let field_values = ["abc", "xyz_but_long_enough_to_hash"];
+    {
+        let mut snapshot = storage.clone().open_snapshot_write();
+        let mut attrs = Vec::new();
+        for val in field_values {
+            let nested_struct_value = StructValue::new(
+                nested_struct_key.clone(),
+                HashMap::from([(0, FieldValue::String(Cow::Borrowed(val)))]),
+            );
+            let outer_struct_value = StructValue::new(
+                struct_key.clone(),
+                HashMap::from([(0, FieldValue::Struct(Cow::Owned(nested_struct_value)))]),
+            );
+            let attr = thing_manager
+                .create_attribute(&mut snapshot, attr_type.clone(), Value::Struct(Cow::Borrowed(&outer_struct_value)))
+                .unwrap();
+            attrs.push(attr);
+        }
+
+        for (val, attr) in std::iter::zip(field_values, attrs.iter()) {
+            let mut attr_by_field_iterator = thing_manager
+                .get_attributes_by_struct_field(
+                    &snapshot,
+                    attr_type.clone(),
+                    vec![0],
```

Comment:
Can we get this ID `0` (presumably this is the field ID?) from the definition instead of hardcoding it?